Patch to fix overwriting productivity gains from research

josephquesnel opened the pull request:
Fixed the overwriting issue by storing a cache of past percentage gains. Fixed a follow up issue with migrating to this version: The updated cache was zeroed out leading to a doubling of the bonus. Since this mod previously overwrote productivity, it just zeroes all mod affected production before applying the 'new' bonus.

Main changes are to:
product_cache.lua - on_production_statistics_may_have_changed function: now properly migrates if flag is set
product_cache.lua - on_production_statistics_may_have_changed function: bonuses from research are kept and are no longer overwritten.
control.lua - added onload func to set migration flag.

Works with initial testing.

Only thing you might want to change is the decimals shown.

josephquesnel commented:
Okay there are still some issues:
mainly, it migrates every time it loads. Fixing this today.

ndamiano commented:
I'll test it out in a little bit, and get it merged shortly thereafter. Thank you for taking this on!

josephquesnel commented:
Im putting in the final fixes and testing now i forgot to consider different migration strategies.
if it works, it should do this:
- if an old mod version is detected, it will just take the max productivity between versions (preserving research gains if any)
- if its an old un-modded save detected, it will Add the productivity which will keep previous productivity research intact.

josephquesnel commented:
Okay I know it looks like a lot of commits but most of them are formatting. Also, no idea why the diff view is so wild.

The control.lua is mostly the same.
The product_cache.lua has a few key changes:
1) the setupStorage function overwrites the storage.items
2) added a function to get all the production technologies and their bonus, mapped similar to storage.items but flattened to recipe.
3) changed the final calculation to include the tech bonus.

Feel free to test, I tried a few different mods including one called too-many-prods which has a productivity research for every item and I think was mentioned in the mod portal discussions. I use a lot of those mods too. Im pretty satisfied with it now.

future improvements:
- There is an issue with the formatting for the percentages, some of them do not truncate at 2 decimal places.
- Could probably remove the string match for productivity, it would be a tiny bit slower but would guarantee compatibility for mods with weird research names
